front/app/components/admin/GraphCards/CommentsByTimeCard/useCommentsByTime/parse.ts

```diff
@@ -0,0 +1,123 @@
+import moment, { Moment } from 'moment';
+
+// utils
+import { timeSeriesParser } from 'components/admin/GraphCards/_utils/timeSeries';
+import { get } from 'utils/helperUtils';
+import { orderBy } from 'lodash-es';
+
+// typings
+import {
+  Response,
+  TimeSeriesResponseRow,
+  TimeSeries,
+  TimeSeriesRow,
+} from './typings';
+import { Translations } from './translations';
+import { IResolution } from 'components/admin/ResolutionControl';
+
+export const getEmptyRow = (date: Moment) => ({
+  date: date.format('YYYY-MM-DD'),
+  comments: 0,
+  total: 0,
+});
+
+const parseRow = (date: Moment, row?: TimeSeriesResponseRow): TimeSeriesRow => {
+  if (!row) return getEmptyRow(date);
+
+  return {
+    total: 0,
+    comments: row.count,
+    date: date.format('YYYY-MM-DD'),
+  };
+};
+
+const getDate = (row: TimeSeriesResponseRow) => {
+  return moment(get(row, 'first_dimension_date_created_date'));
+};
+
+const _parseTimeSeries = timeSeriesParser(getDate, parseRow);
+
+export const parseTimeSeries = (
+  responseTimeSeries: Response['data'][0],
+  startAtMoment: Moment | null | undefined,
+  endAtMoment: Moment | null,
+  resolution: IResolution,
+  total: Response['data'][1]
+): TimeSeries | null => {
+  let timeSeries = _parseTimeSeries(
+    responseTimeSeries,
+    startAtMoment,
+    endAtMoment,
+    resolution
+  );
+  if (!timeSeries) return [];
+  let totalCount = total && total.length > 0 ? total[0]?.count : 0;
+  timeSeries = orderBy(
+    timeSeries,
+    (o: TimeSeriesRow) => {
+      return moment(o.date).format('YYYYMMDD');
+    },
+    ['desc']
+  )
+    .map((row) => {
+      const _totalCount = totalCount;
+      totalCount = (totalCount || 0) - row.comments;
+      return {
+        ...row,
+        total: _totalCount || 0,
+      };
+    })
+    .reverse();
```

Comment:
Why are we reversing this time series (sorting it descending by date), then taking the total (which, judging from the code, should actually never be 0 because otherwise this is all going to be negative) and progressively subtracting the comments from that? Why not skip the whole double reverse stuff and just start from the lowest date, loop over the array in order, and keep adding things to the total instead of doing this double reverse?

Reply:
having the global total, and being the same as the last element in the cumulative serie, im building here the whole cumulative serie. 
so i have to loop the serie starting from the oldest date and substracting each period value to get the total on that point. then reversing again to render the dates on the chart in the correct order.

Reply:
Ah I think I get it- the 'total' is the total over all time, while `responseTimeSeries` is only the data the requested period. So if the whole period is January: 10, February: 10, March: 10, but you only request February and March, you still want the total at the end of march to be 30 (and the cumulative total at the end of february 20). Is that correct?

But in this case I think there's a bug: since you're not applying the date filter to the `commentsByTimeTotal` query in `query.ts`, you will always get the total of the whole period. So if in the example you would request only January and February, this function will now return 30 for the cumulative total at the end of February, which is wrong. So I think you have to use the `endAt` date for the `commentsByTimeTotal` query too.

Also I think it would be cleaner to remove all these `|| 0` checks. We should request the total, and if we get a number back, we should just use that number. If we don't get a number back, we shouldn't be doing any of these calculations. So I would replace

```ts
if (!timeSeries) return [];
let totalCount = total[0]?.count;
```

by

```ts
let totalCount = total[0]?.count;
if (!timeSeries || timeSeries.length === 0 || typeof totalCount !== 'number') return null;
```

Also I think that if there's anything wrong we should prefer returning `null` over an empty array, especially since we usually only check if `isNilOrError` and not for empty arrays- see for example `front/app/components/admin/GraphCards/VotesByTimeCard/Chart.tsx`